controllers/vrg_kubeobjects.go

```diff
@@ -0,0 +1,585 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"errors"
+	"strconv"
+	"time"
+
+	ramen "github.com/ramendr/ramen/api/v1alpha1"
+	"github.com/ramendr/ramen/controllers/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+func s3PathNamePrefix(vrgNamespaceName, vrgName string) string {
+	return types.NamespacedName{Namespace: vrgNamespaceName, Name: vrgName}.String() + "/"
+}
+
+const vrgS3ObjectNameSuffix = "a"
+
+func VrgObjectProtect(objectStorer ObjectStorer, vrg ramen.VolumeReplicationGroup) error {
+	return uploadTypedObject(objectStorer, s3PathNamePrefix(vrg.Namespace, vrg.Name), vrgS3ObjectNameSuffix, vrg)
+}
+
+func VrgObjectUnprotect(objectStorer ObjectStorer, vrg ramen.VolumeReplicationGroup) error {
+	return DeleteTypedObjects(objectStorer, s3PathNamePrefix(vrg.Namespace, vrg.Name), vrgS3ObjectNameSuffix, vrg)
+}
+
+func vrgObjectDownload(objectStorer ObjectStorer, pathName string, vrg *ramen.VolumeReplicationGroup) error {
+	return downloadTypedObject(objectStorer, pathName, vrgS3ObjectNameSuffix, vrg)
+}
+
+func kubeObjectsCaptureInterval(kubeObjectProtectionSpec *ramen.KubeObjectProtectionSpec) time.Duration {
+	if kubeObjectProtectionSpec.CaptureInterval == nil {
+		return ramen.KubeObjectProtectionCaptureIntervalDefault
+	}
+
+	return kubeObjectProtectionSpec.CaptureInterval.Duration
+}
+
+func timeSincePreviousAndUntilNext(previousTime time.Time, interval time.Duration) (time.Duration, time.Duration) {
+	since := time.Since(previousTime)
+
+	return since, interval - since
+}
+
+func kubeObjectsCapturePathNameAndNamePrefix(namespaceName, vrgName string, captureNumber int64) (string, string) {
+	number := strconv.FormatInt(captureNumber, 10)
+
+	return s3PathNamePrefix(namespaceName, vrgName) + "kube-objects/" + number + "/",
+		// TODO fix: may exceed name capacity
+		namespaceName + "--" + vrgName + "--" + number
+}
+
+func kubeObjectsCaptureName(prefix, groupName, s3ProfileName string) string {
+	return prefix + "--" + groupName + "--" + s3ProfileName
+}
+
+func kubeObjectsRecoverNamePrefix(vrgNamespaceName, vrgName string) string {
+	return vrgNamespaceName + "--" + vrgName
+}
+
+func kubeObjectsRecoverName(prefix string, groupNumber int) string {
+	return prefix + "--" + strconv.Itoa(groupNumber)
+}
+
+func (v *VRGInstance) kubeObjectsProtect(result *ctrl.Result) {
+	s3StoreAccessors := v.s3StoreAccessorsGet()
+	if s3StoreAccessors == nil {
+		result.Requeue = true
+
+		return
+	}
+
+	if !v.kubeObjectProtectionDisabled("capture") {
+		v.kubeObjectsCaptureStartOrResumeOrDelay(result, s3StoreAccessors)
+	}
```

Comment:
Don't you need to return here in `else` branch? Well, I guess the question is why do you need to protect the vrg if kubeobject protection is disabled?